File: frontend/src/core/architecture/service-registry.ts
import { Service } from './service';

export class ServiceRegistry {
  private services = new Map<string, Service>();

  public register<T extends Service>(service: T): T {
    if (this.services.has(service.serviceName)) {
      console.warn(`[ServiceRegistry] Service ${service.serviceName} already registered. Overwriting.`);
    }
    this.services.set(service.serviceName, service);
    return service;
  }

  public get<T extends Service>(name: string): T {
    const service = this.services.get(name);
    if (!service) {
      throw new Error(`[ServiceRegistry] Service '${name}' not found.`);
    }
    return service as T;
  }

  public has(name: string): boolean {
    return this.services.has(name);
  }

  public async clear(): Promise<void> {
    for (const service of this.services.values()) {
      if (service.dispose) {
        await service.dispose();
      }
    }
    this.services.clear();
  }
}

File: frontend/Assets/_FLOWSTATE/Core/Architecture/ServiceRegistry.cs
// =============================================================================
// FLOWSTATE — ServiceRegistry.cs
// Central service locator with explicit registration. Avoids singleton sprawl.
// =============================================================================

using System;
using System.Collections.Generic;
using UnityEngine;

namespace Flowstate.Core.Architecture
{
    /// <summary>
    /// Central registry for all game services. Provides controlled access
    /// to services without relying on global singletons.
    ///
    /// Usage:
    ///   ServiceRegistry.Register&lt;IAudioService&gt;(audioService);
    ///   var audio = ServiceRegistry.Get&lt;IAudioService&gt;();
    /// </summary>
    public static class ServiceRegistry
    {
        private static readonly Dictionary<Type, IService> _services = new();
        private static bool _isInitialized;

        /// <summary>
        /// Register a service by its interface type.
        /// </summary>
        public static void Register<T>(T service) where T : class, IService
        {
            var type = typeof(T);
            if (_services.ContainsKey(type))
            {
                Debug.LogWarning($"[ServiceRegistry] Overwriting existing service: {type.Name}");
            }
            _services[type] = service;
        }

        /// <summary>
        /// Retrieve a registered service by its interface type.
        /// </summary>
        public static T Get<T>() where T : class, IService
        {
            var type = typeof(T);
            if (_services.TryGetValue(type, out var service))
            {
                return service as T;
            }

            Debug.LogError($"[ServiceRegistry] Service not found: {type.Name}");
            return null;
        }

        /// <summary>
        /// Check if a service is registered.
        /// </summary>
        public static bool Has<T>() where T : class, IService
        {
            return _services.ContainsKey(typeof(T));
        }

        /// <summary>
        /// Initialize all registered services in registration order.
        /// </summary>
        public static void InitializeAll()
        {
            foreach (var service in _services.Values)
            {
                service.Initialize();
            }
            _isInitialized = true;
        }

        /// <summary>
        /// Shutdown all services and clear the registry.
        /// </summary>
        public static void ShutdownAll()
        {
            foreach (var service in _services.Values)
            {
                service.Shutdown();
            }
            _services.Clear();
            _isInitialized = false;
        }
    }
}
